enhancements/upgrade-readiness.md

```diff
@@ -0,0 +1,543 @@
+apiVersion: operators.coreos.com/v1alpha1
+kind: Probe
+metadata:
+  name: foo-operator.v1.0.0
+  namespace: operators
+spec:
+  specOverride: crdAnnotations # Changes to the CRD's annotations should override the spec. SPO will watch the CRD for changes to the "OLM Supported Condition" annotations and update the spec accordingly.
+  probeResources:
+  - resource: foo.example.com
+    upgradeable: "!Migrating"
+    important: "BadConnectivity || UnhealthyDatabase"
+```
+
+When the`probe` resource above is created, OLM will spin up an SPO pod which collects the information from conditions of `foo.example.com` CRs and communicate it to OLM via the `probe`'s status accordingly. If an operator were to define multiple resources that it communicates its status through, a single SPO pod would monitor all of those resources. The permissions of the probe will match the same permissions of the operator.
+
+>Note: If the probe's spec is nil, SPO will not create a pod and the operator will be expected to update the CRs status itself.
+
+In this example, the SPO will be responsible for maintaining the following fields in the status of the `probe` resource:
+
+```yaml
+status:
+  conditions:
+    upgradeable: # A boolean that represents if the operator may be upgraded - aggregated from all probeResources below.
+  probeResources: # The list of resources the probe will watch and the conditions of those resources.
+```
+
+Consider an upgrade where the a single `Foo` CR exists on cluster:
+
+```yaml
+apiVersion: foo.example.com/v1
+kind: Foo
+metadata:
+  name: foo-example
+  namespace: default
+spec:
+  ...
+  ...
+  ...
+status:
+  conditions:
+  - type: Migrating # Condition 1
+    status: true
+    reason: "CriticalProcess"
+    message: "Migrating old database schema to new format"
+  - type: UnhealthyDatabase # Condition 2
+    status: true
+    reason: "LivenessProbeFailure"
+    message: "The foo-database failed its liveness probe."
+  - type: BadConnectivity # Condition 3
+    status: true
+    reason: "404"
+    message: "Unable to reach the database"
+  - type: SomethingUnrelated # Condition 4
+    status: true
+    reason: "Unmatched"
+    message: "OLM doesn't care about this condition"
+```
+
+When SPO registers this CR, the Probe CR will be updated like so:
+
+```yaml
+apiVersion: operators.coreos.com/v1alpha1
+kind: Probe
+metadata:
+  name: foo-operator.v1.0.0 # Same name as the CSV.
+  namespace: operators
+spec:
+  specOverride: crdAnnotations # When set to `crdAnnotations` OLM will override any changes to the probe's spec to match those defined via annotations on the CRDs described above.
+  probeResources:
+  - name: foo.example.com
+    upgradeable: "!Migrating"
+    important: "BadConnectivity || UnhealthyDatabase"
+status:
+  conditions:
+  - type: Upgradeable # Overall Upgradeable condition
+    status: false
+    reason: "NotUpgradeable"
+    message: "The operator has communicated that the operator is not upgradeable"
+  probeResources:
```

Comment:
this is going to get un-readable once you add three CR's nevermind 100. 

Can these instead become refs to the CR's that are causing the condition to trigger?

````yaml
probeResources:
- name: foo-example
  namespace: default
  kind: Foo
  group: foo.example.com
  reason: "!Migrating"
```

I get the desire to include the entire decision, I am just worried that it will be too much for a user to read.

Reply:
That's fair - I am fine removing the list of conditions and simply including the reason.

Reply:
Given that a single resource may cause multiple OLM Supported Conditions to fire, I suggest making the `reason` field an array of strings.

Reply:
Do we need to explain this at all?  Or can we have something more opaque like [`relatedObjects`][1] that just points at the involved resources, and leaves it up the the included `message` and external docs like this enhancement to spell out the important pieces (e.g. `default/foo-example has UnhealthyDatabase and BadConnectivity` instead of the example's current, generic `The operator has communicated that the operator is not upgradeable`).

[1]: https://github.com/openshift/api/blob/2ea89d203c53704f1fcfeb55c13ededab14fd020/config/v1/types_cluster_operator.go#L49-L54